File: frontend/src/api/attendance.ts
import { api } from "@/lib/api"

export interface AttendanceRecord {
  id: number
  employee_id: number
  employee_name: string
  location_id: number
  location_name: string
  check_in_time: string
  check_out_time: string | null
  is_late: boolean
  late_by_minutes: number
  status: string
  date: string
  distance_from_location_meters: number | null
}

export interface CheckInRequest {
  latitude: number
  longitude: number
}

export interface CheckInResponse extends AttendanceRecord {
  message: string
}

export const attendanceApi = {
  checkin: async (coords: CheckInRequest): Promise<CheckInResponse> => {
    const response = await api.post<CheckInResponse>("/attendance/checkin", coords)
    return response.data
  },

  checkout: async (): Promise<AttendanceRecord> => {
    const response = await api.post<AttendanceRecord>("/attendance/checkout")
    return response.data
  },

  getTodayAttendance: async (): Promise<AttendanceRecord | null> => {
    const response = await api.get<AttendanceRecord | null>("/attendance/today")
    return response.data
  },

  getHistory: async (
    startDate?: string,
    endDate?: string
  ): Promise<AttendanceRecord[]> => {
    const params = new URLSearchParams()
    if (startDate) params.append("start_date", startDate)
    if (endDate) params.append("end_date", endDate)
    const response = await api.get<AttendanceRecord[]>(
      `/attendance/history?${params.toString()}`
    )
    return response.data
  },

  getAllAttendance: async (params?: {
    date?: string
    location_id?: number
    department_id?: number
    employee_id?: number
    page?: number
    page_size?: number
  }): Promise<{
    items: AttendanceRecord[]
    total: number
    page: number
    page_size: number
  }> => {
    const response = await api.get("/attendance/all", { params })
    return response.data
  },
}

File: backend/app/schemas/attendance.py
from typing import Optional
from pydantic import BaseModel, Field
from datetime import datetime, date


class CheckInRequest(BaseModel):
    latitude: float = Field(..., description="Employee's current latitude")
    longitude: float = Field(..., description="Employee's current longitude")


class AttendanceResponse(BaseModel):
    id: int
    employee_id: int
    employee_name: str
    location_id: int
    location_name: str
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None
    is_late: bool
    late_by_minutes: int
    status: str
    date: date
    distance_from_location_meters: Optional[float] = None

    class Config:
        from_attributes = True


class CheckInResponse(AttendanceResponse):
    message: str
    distance_from_location_meters: float


class CheckOutResponse(AttendanceResponse):
    pass


class AttendanceListResponse(BaseModel):
    items: list[AttendanceResponse]
    total: int
    page: int
    page_size: int
